File: 5.02/서지훈_백준_11085_군사이동.py
import sys
from collections import deque


class UnionFind:
    def __init__(self, n):  # parent의 루트 노드를 자기 자신으로 설정합니다.
        self.parent = [i for i in range(n)]  # 자기 자신을 가르키도록 초기화합니다.

    def find(self, target):  # 루트 노드를 탐색합니다.
        if target == self.parent[target]:  # 부모 노드가 자기 자신, 즉 루트 노드이면 리턴합니다.
            return target
        else:  # 재귀적으로 루트 노드를 탐색합니다.
            self.parent[target] = self.find(self.parent[target])
            return self.parent[target]

    def union(self, a, b):
        a = self.find(a)
        b = self.find(b)
        if a >= b:
            self.parent[a] = b
        else:
            self.parent[b] = a

    def is_union(self, a, b):
        if self.find(a) == self.find(b):
            return True
        return False


p, w = map(int, sys.stdin.readline().split())
c, v = map(int, sys.stdin.readline().split())

uf = UnionFind(p)

dq = deque()

for _ in range(w):
    w_start, w_end, w_width = map(int, sys.stdin.readline().split())
    dq.append((w_start, w_end, w_width))

dq = sorted(dq, key=lambda x: -x[2])
print(dq)

for element in dq:
    uf.union(element[0], element[1])
    print(uf.find(c), uf.find(v), element[2])

    if uf.is_union(c, v):
        print(element[2])
        break

File: 5.02/서지훈_백준_11085_군사이동.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class UnionFind{
    private int[] parent;
    private final int parent_length;
    public UnionFind(int N){
        this.parent_length = N+2;
        this.init();
    }
    public void init(){
        parent = new int[parent_length];
        for(int i = 1; i < parent_length; ++i){
            parent[i] = i;
        }
    }
    public void union(int a, int b){
        int rootA = this.find(a);
        int rootB = this.find(b);
        if(rootA >= rootB){
            parent[rootA] = rootB;
        }
        else{
            parent[rootB] = rootA;
        }
    }
    public int find(int target){
        if(target == this.parent[target]) return target;
        int parent_idx = find(parent[target]);
        return parent[target] = parent_idx;
    }
}

class Tuple{
    private final int start, end, width;
    public Tuple(int a, int b, int c){
        this.start = a;
        this.end = b;
        this.width = c;
    }
    public int returnStart(){return this.start;}
    public int returnEnd(){return this.end;}
    public int returnWidth(){return this.width;}
}

public class Main{
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int p = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int c = Integer.parseInt(st.nextToken());
        int v = Integer.parseInt(st.nextToken());

        PriorityQueue<Tuple> pq = new PriorityQueue<>((o1, o2) -> o2.returnWidth() - o1.returnWidth());

        for(int i = 0; i < w; ++i){
            st = new StringTokenizer(br.readLine());
            int w_start = Integer.parseInt(st.nextToken());
            int w_end = Integer.parseInt(st.nextToken());
            int w_width = Integer.parseInt(st.nextToken());
            pq.add(new Tuple(w_start, w_end, w_width));
        }

        UnionFind uf = new UnionFind(p);

        while(!pq.isEmpty()){
            int w_start = pq.peek().returnStart();
            int w_end = pq.peek().returnEnd();
            int w_width = pq.poll().returnWidth();

            uf.union(w_start, w_end);

            if(uf.find(c) == uf.find(v)){
                bw.write(Integer.toString(w_width));
                break;
            }
        }

        bw.flush();
        br.close();
        bw.close();
    }
}